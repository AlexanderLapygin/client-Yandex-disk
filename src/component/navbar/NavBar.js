import React, {Component} from 'react';
import {connect} from 'react-redux'
import {withRouter} from 'react-router-dom'
import {getToken, getinfo, getPath} from "./../../selectors/selectors";
import {appID} from "./../../config";
import {FETCH_INFO_DISK_REQUEST, CLEAR_TOKEN, CLEAR_INFO,CLEAR_DATA,FETCH_RESOURCES_REQUEST} from './../../ActionType';
import Path from './../../component/path/Path'


class NavBar extends Component {
    //если у адреса есть хеш тогда сохраняем токен
    //это будет только в том случае если мы получим урл от яндекса при успешной авторизации
    componentDidMount() {
        if (this.props.location.hash ) {
            const token = /access_token=([^&]+)/.exec(this.props.location.hash)[1];
            const {dispatch} = this.props;
            //выполним запрос на сервер и получим данные по пользователю,
            dispatch({type: FETCH_INFO_DISK_REQUEST, payload: token});
            this.props.history.push('/');
            dispatch({type: FETCH_RESOURCES_REQUEST, payload: this.props.location.pathname});
        }else if(this.props.token && !this.props.info.hasOwnProperty('display_name')){
            const {dispatch} = this.props;
            //выполним запрос на сервер и получим данные по пользователю,
            dispatch({type: FETCH_INFO_DISK_REQUEST, payload: this.props.token});
        }
    }

    signout(){
        const {dispatch} = this.props;
        //очищаем токен в сторе и локалсторедж
        localStorage.setItem('token','');
        dispatch({type: CLEAR_TOKEN});
        dispatch({type:CLEAR_INFO});
        dispatch({type:CLEAR_DATA});
        this.props.history.push('/');
    }

    render() {
        return (
            <div className='container'>
                <nav className="navbar navbar-light bg-light NavBar__main">
                    <ol className="breadcrumb">
                        {this.props.info.display_name ? <Path key={'01'} pathEl={"Файлы"}/>:null}
                        {this.props.history.location.pathname.substr(1).split('/').map((pth, i) => <Path key={i} pathEl={pth}/>)}
                    </ol>
                    {this.props.info.display_name ?
                        <div className='btn-group'>
                            <span className='btn btn-success disabled'>{this.props.info.display_name}</span>
                            <button className='btn btn-success'
                                    onClick={()=>this.signout()}
                            >Выйти</button>
                        </div>
                        :
                        <div>
                        <a className='btn btn-outline-success'
                        href={`https://oauth.yandex.ru/authorize?response_type=token&client_id=${appID}`}
                        target='_self'>Войти</a>
                        </div>}
                </nav>
                <div className='row'>
                    {this.props.children}
                </div>
            </div>
        );
    }
}

const mapStateToProps = state => {
    return ({
        token: getToken(state),
        info: getinfo(state)
    })
};

export default withRouter(connect(mapStateToProps)(NavBar));
