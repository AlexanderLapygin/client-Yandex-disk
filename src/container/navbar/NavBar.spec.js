import React from 'react';
import { shallow } from 'enzyme';
import { NavBar } from './NavBar';

const mockfetchInfoDiskRequest = jest.fn();
const mockfetchResourcesRequest = jest.fn();
const mockclearInfoDisk = jest.fn();
const mockclearResources = jest.fn();
const mockclearToken = jest.fn();
const mockopenModal = jest.fn();

describe('NavBar testing', () => {

// описываем props
  const props = {
    token: '',
    history: {
      location: {
        pathname: '/',
      },
    },
    location: {
      pathname: '/',
    },
    info: {
      display_name: 'Test NavBar',
    },

  };

  describe('NavBar', () => {

    const NavBarContainer = shallow(<NavBar  {...props} />);

    it('snapshot', () => {
      expect(NavBarContainer).toMatchSnapshot();
    });

    describe('NavBar testing functions', () => {

      const nextProps = {
        token: '',
        location: {
          pathname: '/',
          hash: '#access_token=AQAAAAAsVhrAAAU_ByadG6czQUSdgY4RAZ7qiYE&token_type=bearer&expires_in=29850557',

        },
        history:
          {
            location: {
              pathname: '/',
            },
            push: jest.fn(),
          },
        info: {
          display_name: 'Test NavBar',
        },
        fetchInfoDiskRequest: mockfetchInfoDiskRequest,
        fetchResourcesRequest: mockfetchResourcesRequest,
        clearInfoDisk: mockclearInfoDisk,
        clearResources: mockclearResources,
        clearToken: mockclearToken,
        openModal: mockopenModal,
      };

      const NavBarContainer = shallow(<NavBar  {...nextProps} />);

      it('span text', () => {
        expect(NavBarContainer.find('span').text()).toEqual('Test NavBar');
      });

      //проверяем вызов функции mockcreateFolderRequest
      it('function call componentDidMount', () => {
        expect(mockfetchInfoDiskRequest).toHaveBeenCalledTimes(1);
        expect(mockfetchInfoDiskRequest).toHaveBeenCalledTimes(1);
      });

      //проверяем вызов функции mockcloseModal
      it('function signOut', () => {
        NavBarContainer.find('#signOut').simulate('click');
        expect(mockclearResources).toHaveBeenCalled();
        expect(mockclearToken).toHaveBeenCalled();
        expect(mockclearInfoDisk).toHaveBeenCalled();
      });

      it('function create folder', () => {
        NavBarContainer.find('#createFolderBtn').simulate('click');
        expect(mockopenModal).toHaveBeenCalled();
      });
    });
  });
});
