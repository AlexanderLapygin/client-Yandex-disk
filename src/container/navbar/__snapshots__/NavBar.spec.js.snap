// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar testing NavBar snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NavBar
    history={
      Object {
        "location": Object {
          "pathname": "/",
        },
      }
    }
    info={
      Object {
        "display_name": "Test NavBar",
      }
    }
    location={
      Object {
        "pathname": "/",
      }
    }
    token=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <nav
          className="navbar navbar-light bg-light NavBar__main"
        >
          <ol
            className="breadcrumb"
          >
            <withRouter(Connect(Path))
              pathEl="Файлы"
            />
            <withRouter(Connect(Path))
              pathEl=""
            />
          </ol>
          <div
            className="btn-group"
          >
            <span
              className="btn btn-success disabled"
            >
              Test NavBar
            </span>
            <button
              className="btn btn-success"
              id="signOut"
              onClick={[Function]}
            >
              Выйти
            </button>
          </div>
        </nav>,
        <div
          className="row justify-content-center mt-3 mb-3"
        >
          <div
            className="btn-group"
          >
            <button
              className="btn btn-outline-warning"
              id="createFolderBtn"
              onClick={[Function]}
            >
              Создать папку
            </button>
            <label
              className="btn btn-outline-primary mb-0"
            >
              <input
                className="NavBar__input_LoadFile"
                onChange={[Function]}
                type="file"
              />
              Загрузить файл
            </label>
          </div>
        </div>,
        <Connect(Modal)
          location="/"
        />,
        <div
          className="row"
        />,
        <div
          className="row"
        >
          <Connect(MSG) />
        </div>,
      ],
      "className": "container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <ol
              className="breadcrumb"
            >
              <withRouter(Connect(Path))
                pathEl="Файлы"
              />
              <withRouter(Connect(Path))
                pathEl=""
              />
            </ol>,
            <div
              className="btn-group"
            >
              <span
                className="btn btn-success disabled"
              >
                Test NavBar
              </span>
              <button
                className="btn btn-success"
                id="signOut"
                onClick={[Function]}
              >
                Выйти
              </button>
            </div>,
          ],
          "className": "navbar navbar-light bg-light NavBar__main",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <withRouter(Connect(Path))
                  pathEl="Файлы"
                />,
                Array [
                  <withRouter(Connect(Path))
                    pathEl=""
                  />,
                ],
              ],
              "className": "breadcrumb",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "01",
                "nodeType": "function",
                "props": Object {
                  "pathEl": "Файлы",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "0",
                "nodeType": "function",
                "props": Object {
                  "pathEl": "",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "ol",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <span
                  className="btn btn-success disabled"
                >
                  Test NavBar
                </span>,
                <button
                  className="btn btn-success"
                  id="signOut"
                  onClick={[Function]}
                >
                  Выйти
                </button>,
              ],
              "className": "btn-group",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Test NavBar",
                  "className": "btn btn-success disabled",
                },
                "ref": null,
                "rendered": "Test NavBar",
                "type": "span",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Выйти",
                  "className": "btn btn-success",
                  "id": "signOut",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Выйти",
                "type": "button",
              },
            ],
            "type": "div",
          },
        ],
        "type": "nav",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="btn-group"
          >
            <button
              className="btn btn-outline-warning"
              id="createFolderBtn"
              onClick={[Function]}
            >
              Создать папку
            </button>
            <label
              className="btn btn-outline-primary mb-0"
            >
              <input
                className="NavBar__input_LoadFile"
                onChange={[Function]}
                type="file"
              />
              Загрузить файл
            </label>
          </div>,
          "className": "row justify-content-center mt-3 mb-3",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="btn btn-outline-warning"
                id="createFolderBtn"
                onClick={[Function]}
              >
                Создать папку
              </button>,
              <label
                className="btn btn-outline-primary mb-0"
              >
                <input
                  className="NavBar__input_LoadFile"
                  onChange={[Function]}
                  type="file"
                />
                Загрузить файл
              </label>,
            ],
            "className": "btn-group",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Создать папку",
                "className": "btn btn-outline-warning",
                "id": "createFolderBtn",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Создать папку",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <input
                    className="NavBar__input_LoadFile"
                    onChange={[Function]}
                    type="file"
                  />,
                  "Загрузить файл",
                ],
                "className": "btn btn-outline-primary mb-0",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "NavBar__input_LoadFile",
                    "onChange": [Function],
                    "type": "file",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                "Загрузить файл",
              ],
              "type": "label",
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "location": "/",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "className": "row",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Connect(MSG) />,
          "className": "row",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <nav
            className="navbar navbar-light bg-light NavBar__main"
          >
            <ol
              className="breadcrumb"
            >
              <withRouter(Connect(Path))
                pathEl="Файлы"
              />
              <withRouter(Connect(Path))
                pathEl=""
              />
            </ol>
            <div
              className="btn-group"
            >
              <span
                className="btn btn-success disabled"
              >
                Test NavBar
              </span>
              <button
                className="btn btn-success"
                id="signOut"
                onClick={[Function]}
              >
                Выйти
              </button>
            </div>
          </nav>,
          <div
            className="row justify-content-center mt-3 mb-3"
          >
            <div
              className="btn-group"
            >
              <button
                className="btn btn-outline-warning"
                id="createFolderBtn"
                onClick={[Function]}
              >
                Создать папку
              </button>
              <label
                className="btn btn-outline-primary mb-0"
              >
                <input
                  className="NavBar__input_LoadFile"
                  onChange={[Function]}
                  type="file"
                />
                Загрузить файл
              </label>
            </div>
          </div>,
          <Connect(Modal)
            location="/"
          />,
          <div
            className="row"
          />,
          <div
            className="row"
          >
            <Connect(MSG) />
          </div>,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <ol
                className="breadcrumb"
              >
                <withRouter(Connect(Path))
                  pathEl="Файлы"
                />
                <withRouter(Connect(Path))
                  pathEl=""
                />
              </ol>,
              <div
                className="btn-group"
              >
                <span
                  className="btn btn-success disabled"
                >
                  Test NavBar
                </span>
                <button
                  className="btn btn-success"
                  id="signOut"
                  onClick={[Function]}
                >
                  Выйти
                </button>
              </div>,
            ],
            "className": "navbar navbar-light bg-light NavBar__main",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <withRouter(Connect(Path))
                    pathEl="Файлы"
                  />,
                  Array [
                    <withRouter(Connect(Path))
                      pathEl=""
                    />,
                  ],
                ],
                "className": "breadcrumb",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "01",
                  "nodeType": "function",
                  "props": Object {
                    "pathEl": "Файлы",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": "0",
                  "nodeType": "function",
                  "props": Object {
                    "pathEl": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "ol",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <span
                    className="btn btn-success disabled"
                  >
                    Test NavBar
                  </span>,
                  <button
                    className="btn btn-success"
                    id="signOut"
                    onClick={[Function]}
                  >
                    Выйти
                  </button>,
                ],
                "className": "btn-group",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Test NavBar",
                    "className": "btn btn-success disabled",
                  },
                  "ref": null,
                  "rendered": "Test NavBar",
                  "type": "span",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Выйти",
                    "className": "btn btn-success",
                    "id": "signOut",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "Выйти",
                  "type": "button",
                },
              ],
              "type": "div",
            },
          ],
          "type": "nav",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="btn-group"
            >
              <button
                className="btn btn-outline-warning"
                id="createFolderBtn"
                onClick={[Function]}
              >
                Создать папку
              </button>
              <label
                className="btn btn-outline-primary mb-0"
              >
                <input
                  className="NavBar__input_LoadFile"
                  onChange={[Function]}
                  type="file"
                />
                Загрузить файл
              </label>
            </div>,
            "className": "row justify-content-center mt-3 mb-3",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <button
                  className="btn btn-outline-warning"
                  id="createFolderBtn"
                  onClick={[Function]}
                >
                  Создать папку
                </button>,
                <label
                  className="btn btn-outline-primary mb-0"
                >
                  <input
                    className="NavBar__input_LoadFile"
                    onChange={[Function]}
                    type="file"
                  />
                  Загрузить файл
                </label>,
              ],
              "className": "btn-group",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Создать папку",
                  "className": "btn btn-outline-warning",
                  "id": "createFolderBtn",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Создать папку",
                "type": "button",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <input
                      className="NavBar__input_LoadFile"
                      onChange={[Function]}
                      type="file"
                    />,
                    "Загрузить файл",
                  ],
                  "className": "btn btn-outline-primary mb-0",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "className": "NavBar__input_LoadFile",
                      "onChange": [Function],
                      "type": "file",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "input",
                  },
                  "Загрузить файл",
                ],
                "type": "label",
              },
            ],
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "location": "/",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "row",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Connect(MSG) />,
            "className": "row",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
